pkg/resmgr/lib/memory/zones.go

```diff
@@ -154,6 +159,28 @@ func (a *Allocator) zoneFree(zone NodeMask) int64 {
 	return a.zoneCapacity(zone) - a.zoneUsage(zone)
 }
 
+func (a *Allocator) zoneAvailable(zone NodeMask) int64 {
+	available := a.zoneFree(zone)
+	if available <= 0 {
+		return 0
+	}
+
+	// Cap available amount to the smallest available free in any of our ancestors.
+	for _, z := range a.zones {
+		if z.nodes != zone && (z.nodes&zone) == zone {
+			if free := a.zoneFree(z.nodes); free < available {
+				available = free
+			}
+		}
+		if available <= 0 {
```

Comment:
No, it can get below zero in our accounting scheme, because our zones are not disjoint sets of memory nodes, but they are a tree of sets of memory nodes and assignment can happen in any tree node. Now because we don't have granular kernel control about how much memory a container can use from each of the memory nodes it is assigned to, when you assign a container to a zone memset{0,1} we pin it to both and we account for *both* memory zones 0 and 1 the memory limit of the container. If that limit is larger than the memory available in either of the zones, then the zone(s) with less memory than the container limit will have negative amount of available memory. But we cap that in some interface functions to 0 before returning, I'm just not sure if we do it consistently now/yet everywhere, so I don't want to go with a (therefore) risky == 0 test here.